adapter/serviceControl/serviceController.go

```diff
@@ -0,0 +1,215 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package serviceControl
+
+import (
+	"bytes"
+	"fmt"
+	"time"
+
+	"github.com/pborman/uuid"
+	sc "google.golang.org/api/servicecontrol/v1"
+
+	"istio.io/mixer/adapter/serviceControl/config"
+	"istio.io/mixer/pkg/adapter"
+)
+
+type (
+	builder struct {
+		adapter.DefaultBuilder
+
+		createClientFunc
+	}
+
+	aspect struct {
+		serviceName string
+		service     *sc.Service
+		logger      adapter.Logger
+	}
+)
+
+var (
+	name        = "serviceControl"
+	defaultConf = &config.Params{
+		ServiceName: "library-example.sandbox.googleapis.com",
+	}
+)
+
+// Register records the builders exposed by this adapter.
+func Register(r adapter.Registrar) {
+	b := &builder{
+		adapter.NewDefaultBuilder(
+			name,
+			"Check/report to GCP service controller",
+			defaultConf,
+		),
+		createClient,
+	}
+	r.RegisterMetricsBuilder(b)
+	r.RegisterApplicationLogsBuilder(b)
+}
+
+func (*builder) ValidateConfig(c adapter.Config) (ce *adapter.ConfigErrors) {
+	params := c.(*config.Params)
+	if params.ServiceName == "" {
+		ce = ce.Appendf("serviceName", "empty GCP service name")
+	}
+	return
+}
+
+func (b *builder) NewMetricsAspect(env adapter.Env, cfg adapter.Config, metrics map[string]*adapter.MetricDefinition) (adapter.MetricsAspect, error) {
+	return b.newAspect(env, cfg)
+}
+
+func (b *builder) NewApplicationLogsAspect(env adapter.Env, cfg adapter.Config) (adapter.ApplicationLogsAspect, error) {
+	return b.newAspect(env, cfg)
+}
+
+func (b *builder) newAspect(env adapter.Env, cfg adapter.Config) (*aspect, error) {
+	params := cfg.(*config.Params)
+	ss, err := b.createClientFunc(env.Logger())
+	if err != nil {
+		return nil, err
+	}
+	return &aspect{params.ServiceName, ss, env.Logger()}, nil
+}
+
+func (a *aspect) Record(values []adapter.Value) error {
+	b := bytes.NewBufferString("mixer-metric-report-id-")
+	_, err := b.WriteString(uuid.New())
+	if err != nil {
+		return err
+	}
+	rq := a.record(values, time.Now(), b.String())
+	rp, err := a.service.Services.Report(a.serviceName, rq).Do()
+	if err != nil {
+		return err
+	}
+
+	if a.logger.VerbosityLevel(4) {
+		a.logger.Infof("service control metric report operation id %s\nmetrics %v\nresponse %v",
+			rq.Operations[0].OperationId, values, rp)
+	}
+	return nil
+}
+
+func (a *aspect) record(values []adapter.Value, now time.Time, operationID string) *sc.ReportRequest {
+	var vs []*sc.MetricValueSet
+	for _, v := range values {
+		// TODO Only support request count.
+		if v.Definition.Name != "request_count" {
```

Comment:
Is this a requirement on service control side; does it not allow dynamic creation of metrics? In other words, why do we need this check, and what work is required to make it work for arbitrary metrics?